File: src/main/java/mdnet/base/ServerSettings.java
package mdnet.base;

import com.google.gson.annotations.SerializedName;

import java.util.Objects;

public final class ServerSettings {
	@SerializedName("image_server")
	private final String imageServer;
	private final TlsCert tls;
	@SerializedName("latest_build")
	private final int latestBuild;

	public ServerSettings(String imageServer, TlsCert tls, int latestBuild) {
		this.imageServer = Objects.requireNonNull(imageServer);
		this.tls = tls;
		this.latestBuild = latestBuild;
	}

	public String getImageServer() {
		return imageServer;
	}

	public TlsCert getTls() {
		return tls;
	}

	public int getLatestBuild() {
		return latestBuild;
	}

	@Override
	public String toString() {
		return "ServerSettings{" + "imageServer='" + imageServer + '\'' + ", tls=" + tls + ", latestBuild="
				+ latestBuild + '}';
	}

	@Override
	public boolean equals(Object o) {
		if (this == o)
			return true;
		if (o == null || getClass() != o.getClass())
			return false;

		ServerSettings that = (ServerSettings) o;

		if (!imageServer.equals(that.imageServer))
			return false;
		return Objects.equals(tls, that.tls);
	}

	@Override
	public int hashCode() {
		int result = imageServer.hashCode();
		result = 31 * result + (tls != null ? tls.hashCode() : 0);
		return result;
	}

	public static final class TlsCert {
		@SerializedName("created_at")
		private final String createdAt;
		@SerializedName("private_key")
		private final String privateKey;
		private final String certificate;

		public TlsCert(String createdAt, String privateKey, String certificate) {
			this.createdAt = Objects.requireNonNull(createdAt);
			this.privateKey = Objects.requireNonNull(privateKey);
			this.certificate = Objects.requireNonNull(certificate);
		}

		public String getCreatedAt() {
			return createdAt;
		}

		public String getPrivateKey() {
			return privateKey;
		}

		public String getCertificate() {
			return certificate;
		}

		@Override
		public String toString() {
			return "TlsCert{" + "createdAt='" + createdAt + '\'' + ", privateKey='" + "<hidden>" + '\''
					+ ", certificate='" + "<hidden>" + '\'' + '}';
		}

		@Override
		public boolean equals(Object o) {
			if (this == o)
				return true;
			if (o == null || getClass() != o.getClass())
				return false;

			TlsCert tlsCert = (TlsCert) o;

			if (!createdAt.equals(tlsCert.createdAt))
				return false;
			if (!privateKey.equals(tlsCert.privateKey))
				return false;
			return certificate.equals(tlsCert.certificate);
		}

		@Override
		public int hashCode() {
			int result = createdAt.hashCode();
			result = 31 * result + privateKey.hashCode();
			result = 31 * result + certificate.hashCode();
			return result;
		}
	}
}

File: src/main/kotlin/mdnet/base/settings/ServerSettings.kt
/*
Mangadex@Home
Copyright (c) 2020, MangaDex Network
This file is part of MangaDex@Home.

MangaDex@Home is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

MangaDex@Home is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this MangaDex@Home.  If not, see <http://www.gnu.org/licenses/>.
 */
package mdnet.base.settings

import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.databind.annotation.JsonNaming
import dev.afanasev.sekret.Secret

@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy::class)
data class ServerSettings(
    val imageServer: String,
    val latestBuild: Int,
    val url: String,
    val tokenKey: ByteArray,
    val compromised: Boolean,
    val paused: Boolean,
    val forceTokens: Boolean = false,
    val tls: TlsCert?
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ServerSettings

        if (imageServer != other.imageServer) return false
        if (latestBuild != other.latestBuild) return false
        if (url != other.url) return false
        if (!tokenKey.contentEquals(other.tokenKey)) return false
        if (compromised != other.compromised) return false
        if (paused != other.paused) return false
        if (tls != other.tls) return false

        return true
    }

    override fun hashCode(): Int {
        var result = imageServer.hashCode()
        result = 31 * result + latestBuild
        result = 31 * result + url.hashCode()
        result = 31 * result + tokenKey.contentHashCode()
        result = 31 * result + compromised.hashCode()
        result = 31 * result + paused.hashCode()
        result = 31 * result + (tls?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "ServerSettings(imageServer='$imageServer', latestBuild=$latestBuild, url='$url', tokenKey=$tokenKey, compromised=$compromised, paused=$paused, forceTokens=$forceTokens, tls=$tls)"
    }
}

@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy::class)
data class TlsCert(
    val createdAt: String,
    @field:Secret val privateKey: String,
    @field:Secret val certificate: String
)
